### PR TITLE
Get around boost 1.67 boost::filesystem::copy null pointer dereferencing issue

### PR DESCRIPTION
**Pull request by abitmore**
Replaces #42.

Fixes a cli_wallet crash when trying to import private key with `import_key` command.

### CODE REVIEW

**Review by pmconrad — CHANGES_REQUESTED** [review]
As discussed on telegram, boost::system::system_error::what() returns a pointer to a null-terminated  C-string. FC_THROW wraps that pointer in a variant_object for later use, but the string it points to could go out of scope together with the caught exception. Please turn the C-string into a string again.
Same for boost::system::error_code::category().name() (watch out for NULL!).

File: src/filesystem.cpp [hunk 1]
```diff
@@ -245,12 +245,12 @@ namespace fc {
...
          {
-            FC_THROW( "Copy from ${srcfile} to ${dstfile} failed because ${reason}:${value}",
-                      ("srcfile",f)("dstfile",t)("reason",ec.category().name())("value",ec.value()) );
+            FC_THROW( "Copy from ${srcfile} to ${dstfile} failed because ${code} : ${message}",
+                      ("srcfile",f)("dstfile",t)("code",ec.value())("message",ec.message()) );
          }
      } catch ( boost::system::system_error& e ) {
      	FC_THROW( "Copy from ${srcfile} to ${dstfile} failed because ${reason}",
-	         ("srcfile",f)("dstfile",t)("reason",e.what() ) );
+	         ("srcfile",f)("dstfile",t)("reason",std::string(e.what()) ) );
```
Comment (jmjatlanta) [inline src/filesystem.cpp]: I do not want to get into a long discussion about this, but I do not see how the c-string returned by e.what() has an opportunity to go out of scope before FC_THROW returns (barring threads, which I do not think are in play here). Therefore, I see the std::string(e.what()) as useless.

Now, I can see how using c-strings can cause scope issues before we enter the catch (see https://pastebin.com/nkKL8W9n ). But wrapping e.what() in std::string is closing the barn door after the cattle have escaped.

Someone please help me see the error in my thinking.

**Comment by jmjatlanta** [comment]
I have determined that the boost bug reported at https://svn.boost.org/trac10/ticket/13585 has been there since at least 1.55. Evidently newer compilers are bringing this logic error to light. 

In a nutshell: 
``boost::filesystem::operations.cpp: void copy(const path& from, const path& to, system::error_code* ec) ``
dereferences ec when ec could be null. That is undefined behavior, and evidently now, a nearby crash.

Passing a non-null ec (as this PR does) avoids the issue.

File: src/filesystem.cpp (hunk 1 above)
Comment (pmconrad) [reply src/filesystem.cpp]: FC_THROW "returns" by throwing a **new** exception. In my understanding, this makes the flow of control leave the catch block, and e goes out of scope.
The new exception thrown by FC_THROW wraps the original pointer, which remains dangling around after e has been destroyed.

**Review by pmconrad — APPROVED** [review]

**Pull request merged**